apis/v1alpha2/validation/httproute_test.go

```diff
@@ -553,3 +553,89 @@ func TestValidateServicePort(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateHTTPRouteTypeMatchesField(t *testing.T) {
+	tests := []struct {
+		name        string
+		routeFilter *gatewayv1a2.HTTPRouteFilter
+		errCount    int
+	}{
+		{
+			name: "valid HTTPRouteFilterRequestHeaderModifier type filter with matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:                  gatewayv1a2.HTTPRouteFilterRequestHeaderModifier,
+				RequestHeaderModifier: &gatewayv1a2.HTTPRequestHeaderFilter{},
+			},
+			errCount: 0,
+		},
+		{
+			name: "invalid HTTPRouteFilterRequestHeaderModifier type filter with non-matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:          gatewayv1a2.HTTPRouteFilterRequestHeaderModifier,
+				RequestMirror: &gatewayv1a2.HTTPRequestMirrorFilter{},
+			},
+			errCount: 1,
+		},
+		{
+			name: "valid HTTPRouteFilterRequestMirror type filter with matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:          gatewayv1a2.HTTPRouteFilterRequestMirror,
+				RequestMirror: &gatewayv1a2.HTTPRequestMirrorFilter{},
+			},
+			errCount: 0,
+		},
+		{
+			name: "invalid HTTPRouteFilterRequestMirror type filter with non-matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:                  gatewayv1a2.HTTPRouteFilterRequestMirror,
+				RequestHeaderModifier: &gatewayv1a2.HTTPRequestHeaderFilter{},
+			},
+			errCount: 1,
+		},
+		{
+			name: "valid HTTPRouteFilterRequestRedirect type filter with matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:            gatewayv1a2.HTTPRouteFilterRequestRedirect,
+				RequestRedirect: &gatewayv1a2.HTTPRequestRedirectFilter{},
+			},
+			errCount: 0,
+		},
+		{
+			name: "invalid HTTPRouteFilterRequestRedirect type filter with non-matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:          gatewayv1a2.HTTPRouteFilterRequestRedirect,
+				RequestMirror: &gatewayv1a2.HTTPRequestMirrorFilter{},
+			},
+			errCount: 1,
+		},
+		{
+			name: "valid HTTPRouteFilterExtensionRef type filter with matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:         gatewayv1a2.HTTPRouteFilterExtensionRef,
+				ExtensionRef: &gatewayv1a2.LocalObjectReference{},
+			},
+			errCount: 0,
+		},
+		{
+			name: "invalid HTTPRouteFilterExtensionRef type filter with non-matching field",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{
+				Type:          gatewayv1a2.HTTPRouteFilterExtensionRef,
+				RequestMirror: &gatewayv1a2.HTTPRequestMirrorFilter{},
+			},
+			errCount: 1,
+		},
+		{
+			name:        "invalid nil type filter",
+			routeFilter: &gatewayv1a2.HTTPRouteFilter{},
+			errCount:    1,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			errs := validateHTTPRouteFilterTypeMatchesValues(tc.routeFilter, field.NewPath("spec").Child("rules"))
```

Comment:
I think the actual path you'd want here would be something like `spec.rules[0].filters[0]`, but since this is just a test, an empty value is probably sufficient:

```suggestion
			errs := validateHTTPRouteFilterTypeMatchesValues(tc.routeFilter, field.NewPath(""))
```